crates/dbsp/src/trace/ord/fallback/zset.rs

```diff
@@ -0,0 +1,1006 @@
+
+    fn num_entries_shallow(&self) -> usize {
+        match &self.inner {
+            Inner::File(file) => file.num_entries_shallow(),
+            Inner::Vec(vec) => vec.num_entries_shallow(),
+        }
+    }
+
+    fn num_entries_deep(&self) -> usize {
+        match &self.inner {
+            Inner::File(file) => file.num_entries_deep(),
+            Inner::Vec(vec) => vec.num_entries_deep(),
+        }
+    }
+}
+
+impl<K, R> NegByRef for FallbackZSet<K, R>
+where
+    K: DataTrait + ?Sized,
+    R: WeightTraitTyped + ?Sized,
+    R::Type: DBWeight + ZRingValue + NegByRef + Erase<R>,
+{
+    #[inline]
+    fn neg_by_ref(&self) -> Self {
+        Self {
+            factories: self.factories.clone(),
+            inner: match &self.inner {
+                Inner::File(file) => Inner::File(file.neg_by_ref()),
+                Inner::Vec(vec) => Inner::Vec(vec.neg_by_ref()),
+            },
+        }
+    }
+}
+
+impl<K, R> Neg for FallbackZSet<K, R>
+where
+    K: DataTrait + ?Sized,
+    R: WeightTraitTyped + ?Sized,
+    R::Type: DBWeight + ZRingValue + NegByRef + Erase<R>,
+{
+    type Output = Self;
+
+    #[inline]
+    fn neg(self) -> Self {
+        self.neg_by_ref()
+    }
+}
+
+impl<K, R> Add<Self> for FallbackZSet<K, R>
+where
+    K: DataTrait + ?Sized,
+    R: WeightTrait + ?Sized,
+{
+    type Output = Self;
+    #[inline]
+
+    fn add(self, rhs: Self) -> Self::Output {
+        if self.is_empty() {
+            rhs
+        } else if rhs.is_empty() {
+            self
+        } else {
+            self.merge(&rhs)
+        }
+    }
+}
+
+impl<K, R> AddAssign<Self> for FallbackZSet<K, R>
+where
+    K: DataTrait + ?Sized,
+    R: WeightTrait + ?Sized,
+{
+    #[inline]
+    fn add_assign(&mut self, rhs: Self) {
+        if !rhs.is_empty() {
+            *self = self.merge(&rhs);
+        }
+    }
+}
+
+impl<K, R> AddAssignByRef for FallbackZSet<K, R>
+where
+    K: DataTrait + ?Sized,
+    R: WeightTrait + ?Sized,
+{
+    #[inline]
+    fn add_assign_by_ref(&mut self, rhs: &Self) {
+        if !rhs.is_empty() {
+            *self = self.merge(rhs);
+        }
+    }
+}
+
+impl<K, R> AddByRef for FallbackZSet<K, R>
+where
+    K: DataTrait + ?Sized,
+    R: WeightTrait + ?Sized,
+{
+    #[inline]
+    fn add_by_ref(&self, rhs: &Self) -> Self {
+        self.merge(rhs)
+    }
+}
+
+impl<K, R> BatchReader for FallbackZSet<K, R>
+where
+    K: DataTrait + ?Sized,
+    R: WeightTrait + ?Sized,
+{
+    type Factories = FallbackZSetFactories<K, R>;
+    type Key = K;
+    type Val = DynUnit;
+    type Time = ();
+    type R = R;
+    type Cursor<'s> = FallbackZSetCursor<'s, K, R>;
+
+    fn factories(&self) -> Self::Factories {
+        self.factories.clone()
+    }
+
+    #[inline]
+    fn cursor(&self) -> Self::Cursor<'_> {
+        FallbackZSetCursor::new(self)
+    }
+
+    #[inline]
+    fn key_count(&self) -> usize {
+        match &self.inner {
+            Inner::File(file) => file.key_count(),
```

Comment:
hand-built dynamic dispatch

Reply:
It would be better to use `dyn` but I was only able to do that for cursors because our other traits are not object-safe.